charms/worker/k8s/src/upgrade.py

```diff
@@ -71,13 +106,109 @@ def pre_upgrade_check(self) -> None:
                 cause=f"Pods not running in namespace(s): {failing_pods}",
                 resolution="Check the logs for the failing pods.",
             )
 
+    def _verify_worker_version(self) -> bool:
+        """Verify the worker version.
+
+        Returns:
+            True if all the worker versions meet the requirements, otherwise False.
+        """
+        worker_version = self.charm.get_worker_versions()
+        if not worker_version:
+            return True
+        dependency_model: DependencyModel = getattr(self.dependency_model, "k8s_service")
+
+        incompatible = {
+            version: units
+            for version, units in worker_version.items()
+            if not verify_requirements(
+                version=version, requirement=dependency_model.dependencies["k8s-worker"]
+            )
+        }
+
+        if incompatible:
+            units_str = "\n".join(
+                f"[{v}]: {', '.join(u.name for u in units)}" for v, units in incompatible.items()
+            )
+            log.error(
+                "k8s worker charm version requirements not met. Incompatible units: %s", units_str
+            )
+            return False
+
+        return True
+
+    def _perform_upgrade(self, services: List[str]) -> None:
+        """Perform the upgrade.
+
+        Args:
+            services: The services to stop and start during the upgrade.
+        """
+        status.add(ops.MaintenanceStatus("Stopping k8s Services."))
+        stop(SNAP_NAME, services=services)
+        status.add(ops.MaintenanceStatus("Upgrading the k8s snap."))
+        snap_management(self.charm)
+        status.add(ops.MaintenanceStatus("Restarting k8s Services."))
+        start(SNAP_NAME, services=services)
+
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        """Handle the upgrade granted event.
+
+        Args:
+            event: The UpgradeGrantedEvent instance.
+        """
+        with status.context(self.charm.unit, exit_status=ops.ActiveStatus("Ready")):
+            self._upgrade(event)
+
+    def _upgrade(self, event: Union[ops.EventBase, ops.HookEvent]) -> None:
+        """Upgrade the snap workload."""
+        trigger = reschedule.PeriodicEvent(self.charm)
+        current_version, _ = snap_version("k8s")
+
+        status.add(ops.MaintenanceStatus("Verifying the cluster is ready for an upgrade."))
+        if not current_version:
+            log.error("Failed to get the version of the k8s snap.")
+            self.set_unit_failed(cause="Failed to get the version of the k8s snap.")
+            status.add(ops.BlockedStatus("Failed to get the version of the k8s snap."))
+            return
+
+        status.add(ops.MaintenanceStatus("Upgrading the charm."))
+
+        if self.charm.lead_control_plane:
+            if not self._verify_worker_version():
+                self.set_unit_failed(
+                    cause="The k8s worker charm version does not meet the requirements."
```

Comment:
I'm a bit confused as to why verify_worker_version is singular but internally it calls get_worker_versions which is plural.

Reply:
Originally, the method verified the consensus version for the workers. But now, we have changed that in order to accommodate multiple `k8s-worker` applications in the model. Context: https://github.com/canonical/k8s-operator/pull/200#discussion_r1866149071